core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1524,6 +1524,139 @@ public static int multiply(int b0, int b1) {
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
+    return overflow ? null : ans.longValue();
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double and integer values. */
+  public static @Nullable Double safeMultiply(double b0, int b1) {
+    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(b1));
+    boolean overflow = isOverflow(ans, BigDecimal.valueOf(Double.MAX_VALUE),
+        BigDecimal.valueOf(-Double.MAX_VALUE));
+    return overflow ? null : ans.doubleValue();
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double values. */
+  public static @Nullable Double safeMultiply(double b0, double b1) {
+    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(b1));
+    boolean overflow = isOverflow(ans, BigDecimal.valueOf(Double.MAX_VALUE),
+        BigDecimal.valueOf(-Double.MAX_VALUE));
+    return overflow ? null : ans.doubleValue();
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double and BigDecimal values. */
+  public static @Nullable BigDecimal safeMultiply(double b0, BigDecimal b1) {
+    BigDecimal ans = multiply(BigDecimal.valueOf(b0), b1);
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Float.MAX_VALUE), BigDecimal.valueOf(-Float.MAX_VALUE));
+    return overflow ? null : ans;
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to double and long values. */
+  public static @Nullable Long safeMultiply(double b0, long b1) {
+    BigDecimal ans = multiply(BigDecimal.valueOf(b0), BigDecimal.valueOf(1));
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
+    return overflow ? null : ans.longValue();
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to BigDecimal and integer values. */
+  public static @Nullable BigDecimal safeMultiply(BigDecimal b0, int b1) {
+    BigDecimal ans = multiply(b0, BigDecimal.valueOf(b1));
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Float.MAX_VALUE), BigDecimal.valueOf(-Float.MAX_VALUE));
+    return overflow ? null : ans;
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to BigDecimal and double values. */
+  public static @Nullable BigDecimal safeMultiply(BigDecimal b0, double b1) {
+    BigDecimal ans = multiply(b0, BigDecimal.valueOf(b1));
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Float.MAX_VALUE), BigDecimal.valueOf(-Float.MAX_VALUE));
+    return overflow ? null : ans;
+  }
+
+  /** SQL <code>SAFE_MULTIPLY</code> function applied to BigDecimal values. */
+  public static @Nullable BigDecimal safeMultiply(BigDecimal b0, BigDecimal b1) {
+    BigDecimal ans = multiply(b0, b1);
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Float.MAX_VALUE), BigDecimal.valueOf(-Float.MAX_VALUE));
```

Comment:
I'm also surprised that you use  `Float` (rather than `Double`). Each SQL type should use the overflow checking for its corresponding Java data type: SMALLINT = short, INTEGER = int, BIGINT = long, REAL = float, DOUBLE = double. (`TINYINT` is tricky because it is signed whereas `byte` is unsigned.)

Note that BigQuery `FLOAT64` matches SQL `DOUBLE` and Java `double`.

I think you should have a version of `safeMultiply` for each java type, e.g. `(short, short)`, `(int, int)`, `(long, long)`, `(float, float)`, `(double, double)`. I don't think you need `BigDecimal`.

For `int` and `long` call `Math.multiplyExact` and catch the exception.

For `short`, call the `int` version and check whether the result is in range.

For `float` and `double` I don't know how you detect overflow. Maybe do ordinary multiplication for now.

Reply:
I agree with both of you, `Double` makes more sense than `Float`. I've changed it.